# 2.1. Data Structures

Module M-227-04: Programming for Data Analytics

Instructor: Nadežda Zeņina

Previous lesson review

## https://ej.uz/z4p4f

## Important Data Structures

* __Tuple__ is a fixed-length immutable sequence of Python objects, (eg., screen resolution (1920, 1080))
* __List__ is a variable-length ordered mutable sequence of Python objects, (eg., shopping items ["milk", "bread", "coffee"])
* __Set__ is an mutable unordered collection of unique Python objects, (eg., programming languages {"Python", "Java", "C++"})
* __Dictionary__ stores a mutable collection of key-value pairs, where key and value are Python objects, (eg., product data {"name": "Laptop", "price": 900}).

In [ ]:
tuple1 = (1, 2, 3)
print("Tuple ", tuple1)
list1 = [1, 2, 3]
print("List ", list1)
set1 = {1, 2, 3}
print("Set ", set1)
dict1 = {1:"One", 2:"Two", 3:"Three"}
print("Dictionary ", dict1)

Tuple  (1, 2, 3)
List  [1, 2, 3]
Set  {1, 2, 3}
Dictionary  {1: 'One', 2: 'Two', 3: 'Three'}


| Structure | Ordered | Mutable | Duplicates | Access by |
|---|---|---|---|---|
| Tuple | yes | no | yes | position / index |
| List | yes | yes | yes | position / index |
| Set | no | yes | no | membership, not index |
| Dictionary | insertion order | yes | keys: no, values: yes | key |

## Tuples vs. Lists

Similar structures:

In [1]:
tuple1 = (1, 2, 3)
print(tuple1)
print(type(tuple1))
print("Third element = ", tuple1[2])

(1, 2, 3)
<class 'tuple'>
Third element =  3


In [2]:
list1 = [1, 2, 3]
print(list1)
print(type(list1))
print("Third element = ", list1[2])

[1, 2, 3]
<class 'list'>
Third element =  3


## Indexing
Lists, tuples and strings are sequences: every element has a position (index).

* Indexing starts at 0
* Negative indices count from the end: -1 is the last element
* len(s) returns the number of elements

In [5]:
cities = ["Riga", "Daugavpils", "Liepaja", "Jelgava", "Jurmala", "Ventspils", "Rezekne"]

print(len(cities))      # number of elements
print(cities[0])        # first element
print(cities[2])        # third element
print(cities[-1])       # last element
print(cities[-2])       # second from the end

# ERROR: list index out of range
# print(cities[7])

7
Riga
Liepaja
Rezekne
Ventspils


## Slicing

sequence[start:stop:step]

* start is included, stop is excluded
* every part is optional: default start = beginning, stop = end, step = 1
* a slice returns a new object of the same type

In [33]:
cities = ["Riga", "Daugavpils", "Liepaja", "Jelgava", "Jurmala", "Ventspils", "Rezekne"]

print(cities[1:4])     # elements 1, 2, 3
print(cities[:3])      # first three
print(cities[3:])      # from index 3 to the end
print(cities[-2:])     # last two
print(cities[::2])     # every second element
print(cities[::-1])    # reversed copy
print(cities[0:5:2])

['Daugavpils', 'Liepaja', 'Jelgava']
['Riga', 'Daugavpils', 'Liepaja']
['Jelgava', 'Jurmala', 'Ventspils', 'Rezekne']
['Ventspils', 'Rezekne']
['Riga', 'Liepaja', 'Jurmala', 'Rezekne']
['Rezekne', 'Ventspils', 'Jurmala', 'Jelgava', 'Liepaja', 'Daugavpils', 'Riga']
['Riga', 'Liepaja', 'Jurmala']


In [15]:
# The same syntax works on a tuple and on a string
rgb = (255, 128, 0)
print(rgb[0], rgb[1:])

course = "Programming for Data Analytics"
print(course[0], course[-1], course[:11])

255 (128, 0)
P s Programming


## Operations common to lists and tuples

| Operation | Meaning |
|---|---|
| `x in s`, `x not in s` | membership test |
| `s + t` | concatenation, creates a new object |
| `s * n` | repetition |
| `len(s)`, `min(s)`, `max(s)`, `sum(s)` | size and aggregate operations |
| `s.index(x)` | position of the first `x` |
| `s.count(x)` | number of occurrences of `x` |
| `sorted(s)` | returns a new sorted list |

In [39]:
# cities = ["Riga", "Daugavpils", "Liepaja", "Jelgava", "Jurmala", "Ventspils", "Rezekne"]

print("Jelgava" in cities)
print("Riga" not in cities)
print(cities[0] + cities[1])
print(("a", "b") * 3)
print(cities.index("Liepaja"))
print((cities).count('Riga'))

True
False
RigaDaugavpils
('a', 'b', 'a', 'b', 'a', 'b')
2
1


*temps* holds the daily temperature from Monday to Sunday. Using indexing and slicing only, print:

* Monday's and Sunday's temperature
* the working days (Mon–Fri)
* the weekend (Sat–Sun)
* every second day, starting from Monday
* the whole week in reverse order

In [40]:
temps = [12.5, 14.1, 9.8, 7.3, 11.0, 15.6, 13.2]
print(temps[0], temps[-1])
print(temps[1:5])
print(temps[5:])
print(temps[::2])
print(temps[::-1])

12.5 13.2
[14.1, 9.8, 7.3, 11.0]
[15.6, 13.2]
[12.5, 9.8, 11.0, 13.2]
[13.2, 15.6, 11.0, 7.3, 9.8, 14.1, 12.5]


### Lists: adding elements
| Method | What it does |
|---|---|
| `lst.append(x)` | adds **one** element to the end |
| `lst.insert(i, x)` | inserts `x` before position `i` |
| `lst.extend(seq)` | adds **all** elements of `seq` to the end |
| `lst + other` | builds a **new** list, originals unchanged |

In [50]:
stops = ["Central Station", "Old Town", "University"]

stops.append("Airport")
print(stops)

stops.insert(1, "Market")
print(stops)

stops.extend(["Harbour", "Zoo"])
print(stops)

more = stops + ["Depot"]           # a NEW list, stops is unchanged
print(more)
print(stops)

['Central Station', 'Old Town', 'University', 'Airport']
['Central Station', 'Market', 'Old Town', 'University', 'Airport']
['Central Station', 'Market', 'Old Town', 'University', 'Airport', 'Harbour', 'Zoo']
['Central Station', 'Market', 'Old Town', 'University', 'Airport', 'Harbour', 'Zoo', 'Depot']
['Central Station', 'Market', 'Old Town', 'University', 'Airport', 'Harbour', 'Zoo']


In [42]:
# Typical mistake: append vs extend
a = [1, 2, 3]
a.append([4, 5])
print(a, len(a))

b = [1, 2, 3]
b.extend([4, 5])
print(b, len(b))

[1, 2, 3, [4, 5]] 4
[1, 2, 3, 4, 5] 5


## Lists: removing elements

| Method | Removes by | Returns |
|---|---|---|
| `lst.remove(x)` | **value** (first match); `ValueError` if missing | nothing |
| `lst.pop()` | position: the last element | the removed element |
| `lst.pop(i)` | position `i` | the removed element |
| `del lst[i]` | position | nothing |
| `lst.clear()` | everything | nothing |

In [44]:
stops = ["Central Station", "Market", "Old Town", "University", "Airport", "Harbour", "Zoo"]

stops.remove("Market")
print(stops)

last = stops.pop()
print(last, stops)

second = stops.pop(1)
print(second, stops)

del stops[0]
print(stops)

stops.clear()
print(stops)

# ERROR: list.remove(x): x not in list
# [1, 2, 3].remove(10)

['Central Station', 'Old Town', 'University', 'Airport', 'Harbour', 'Zoo']
Zoo ['Central Station', 'Old Town', 'University', 'Airport', 'Harbour']
Old Town ['Central Station', 'University', 'Airport', 'Harbour']
['University', 'Airport', 'Harbour']
[]


## Lists: summarising

Built-in functions work on the whole container, no loop needed.

In [45]:
grades = [7, 9, 4, 10, 6]

print("Count  :", len(grades))
print("Sum    :", sum(grades))
print("Min    :", min(grades))
print("Max    :", max(grades))
print("Average:", sum(grades) / len(grades))
print("Top 3  :", sorted(grades, reverse=True)[:3])

Count  : 5
Sum    : 36
Min    : 4
Max    : 10
Average: 7.2
Top 3  : [10, 9, 7]


## Start from the list of grades below and do the steps in this order:

* add the grade 5 to the end
* insert the grade 10 at the beginning
* remove the first 4
* take out the last grade with pop() and print it
* print how many times 9 occurs
* print the grades sorted from highest to lowest without changing grades
* print the three best grades
* print the average, rounded to 2 decimals (round(x, 2))


In [46]:
grades = [7, 9, 4, 10, 6, 8, 4, 9]

In [48]:
# 1
grades.append(5)

# 2
grades.insert(0, 10)

# 3
grades.remove(4)

# 4
last_grade = grades.pop()
print("Removed grade:", last_grade)

# 5
print("Number of 9s:", grades.count(9))

# 6
sorted_grades = sorted(grades, reverse=True)
print("Sorted grades:", sorted_grades)

# 7
print("Top 3 grades:", sorted_grades[:3])

# 8
average = sum(grades) / len(grades)
print("Average:", round(average, 2))

# Check that the original grades list was not sorted
print("Grades:", grades)

Removed grade: 5
Number of 9s: 2
Sorted grades: [10, 10, 10, 9, 9, 8, 7, 6]
Top 3 grades: [10, 10, 10]
Average: 8.62
Grades: [10, 10, 7, 9, 10, 6, 8, 9]


## Key differences:

* Tuples use less memory than lists

In [51]:
import sys

print(sys.getsizeof(tuple1))
print(sys.getsizeof(list1))
# list is mutable and Python keeps some extra space so new elements can be added efficiently.
# A tuple cannot change, so it usually needs less memory.

64
88


* Tuples are used by reference, lists are copied

In [54]:
print(tuple1)
print(list1)

tuple2 = tuple(tuple1)
print("The same tuple? ", tuple2 is tuple1)

list2 = list(list1)
print("The same list? ", list2 is list1)

(1, 2, 3)
[1, 2, 3]
The same tuple?  True
The same list?  False


* Tuples construction is faster than lists

In [55]:
import timeit
print(timeit.timeit("(1,2,3,4,5)",number=1000))
print(timeit.timeit("[1,2,3,4,5]",number=1000))

1.3719999969907803e-05
4.437000006873859e-05


* Tuples are hashable (will explain later)
* Tuples are immutable, e.g., can not add/delete element, can not sort tuples

In [56]:
list1[0]='First'
print("Updated list = ", list1)

Updated list =  ['First', 2, 3]


In [57]:
tuple1[0]='First' #error!
print("Updated tuple = ", tuple1)

TypeError: 'tuple' object does not support item assignment

## "Changing" a tuple = building a new one
* One-element tuple needs a trailing comma: (...,)
* The  tuple methods: count and index

In [60]:
print(tuple1)
tuple_updated = tuple1 + ("First",)
print(tuple_updated)

(1, 2, 3)
(1, 2, 3, 'First')


## Packing and unpacking
* Packing: several values → one tuple
* Unpacking: one tuple → several variables (the number of variables must match)



In [67]:

point = (56.9496, 24.1052, "Riga")

lat, lon, city = point
print(f"city {city}: latitude {lat}, longitude {lon}")

first, *rest = point
print(first, rest)


city Riga: latitude 56.9496, longitude 24.1052
56.9496 [24.1052, 'Riga']


### Tuples: good examples

In [ ]:
weekdays = ('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday','Saturday', 'Sunday')
print(weekdays)
rigaCoords = (56.9677, 24.1056)
print(rigaCoords)
rgbColor = (255, 255, 255) #white
print(rgbColor)


('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday')
(56.9677, 24.1056)
(255, 255, 255)


Tuples: questionable usage

In [ ]:
students = ('Alexey', 'Anar', 'Erika', 'Jevgenijs', 'Janis', 'Vjaceslav', 'Nikita')
print(students)


('Alexey', 'Anar', 'Erika', 'Jevgenijs', 'Janis', 'Vjaceslav', 'Nikita')


## Tuples and strings

Data usually arrives as text.

| Direction | Tool |
|---|---|
| string → list of strings | `text.split(sep)` |
| list / tuple of strings → string | `sep.join(seq)` |
| list ↔ tuple | `tuple(lst)`, `list(tpl)` |
| any object → its text form | `str(obj)` or an f-string |

- `split()` always returns a **list** — convert with `tuple()` if needed
- `join()` works only if **all** elements are strings
- A string and a tuple cannot be added directly with `+`

In [70]:
line = "Riga,Tallinn,Vilnius"

parts = line.split(",")
print(parts, type(parts))

as_tuple = tuple(parts)
print(as_tuple, type(as_tuple))


['Riga', 'Tallinn', 'Vilnius'] <class 'list'>
('Riga', 'Tallinn', 'Vilnius') <class 'tuple'>


In [71]:
words = ("Python", "for", "Data", "Analytics")
print(" ".join(words))
print("-".join(words))

capitals_t = ("Riga", "Tallinn", "Vilnius")

# ERROR: can only concatenate str (not "tuple") to str
# print("Baltic capitals: " + capitals_t)

print("Baltic capitals: " + ", ".join(capitals_t))
print("Baltic capitals: " + str(capitals_t))
print(f"Baltic capitals: {capitals_t}")

Python for Data Analytics
Python-for-Data-Analytics
Baltic capitals: Riga, Tallinn, Vilnius
Baltic capitals: ('Riga', 'Tallinn', 'Vilnius')
Baltic capitals: ('Riga', 'Tallinn', 'Vilnius')


### tuples -> Colab
* Unpack office into three variables and print a sentence such as TSI is located at 56.9279, 24.1880.
* Build a new tuple office_full that also contains the country "Latvia" as the last element.
* Swap the values of p and q in one line.


In [78]:
office = ("TSI", 56.9279, 24.1880)
p, q = "left", "right"
scores = (72, 55, 91, 68)

office = ("TSI", 56.9279, 24.1880)
p, q = "left", "right"
scores = (72, 55, 91, 68)

x = [1, 2, 3]
y = x
z = x[:]

# 1
name, lat, lon = office
print(f"{name} is located at {lat}, {lon}")

# 2
office_full = office + ("Latvia",)
print(office_full)

# 3
p, q = q, p
print(p, q)




TSI is located at 56.9279, 24.188
('TSI', 56.9279, 24.188, 'Latvia')
right left


## Lists vs. Sets

Similar structures, but no duplicates

In [79]:
list1 = [1, 2, 3, 3, 3, 3]
print(list1)
print(type(list1))

[1, 2, 3, 3, 3, 3]
<class 'list'>


In [ ]:
set1 = {1, 2, 3, 3, 3, 3}
print(set1)
print(type(set1))

{1, 2, 3}
<class 'set'>


In [80]:
# From a list to a set; the empty set
visits = ["Riga", "Liepaja", "Riga", "Cesis", "Riga", "Liepaja"]
print(set(visits))

empty_set = set()              # NOT {} - that is an empty dictionary
print(type(empty_set), type({}))

{'Cesis', 'Liepaja', 'Riga'}
<class 'set'> <class 'dict'>


* Sets are faster than lists for checking presence of object

In [81]:
print(5 in [1,2,3,4,5])
print(5 in {1,2,3,4,5})
import timeit
print("Time for checking in list = ", timeit.timeit("5 in [1,2,3,4,5]",number=1000))
print("Time for checking in set = ",timeit.timeit("5 in {1,2,3,4,5}",number=1000))

True
True
Time for checking in list =  4.650000028050272e-05
Time for checking in set =  2.5608999749238137e-05


* Element in sets can not be accessed by index

In [82]:
print(set1[0])

NameError: name 'set1' is not defined

* Set operations

In [ ]:
set2 = {2, 3, 5}
set3 = {2, 4, 6}
print(f"Instersection of {set2} and {set3} is {set2.intersection(set3)}")
print(f"Union of {set2} and {set3} is {set2.union(set3)}")

Instersection of {2, 3, 5} and {2, 4, 6} is {2}
Union of {2, 3, 5} and {2, 4, 6} is {2, 3, 4, 5, 6}


## Set operations: the full picture

| Operation | Method | Operator | Result |
|---|---|---|---|
| Union | `A.union(B)` | `A \| B` | in A **or** B |
| Intersection | `A.intersection(B)` | `A & B` | in A **and** B |
| Difference | `A.difference(B)` | `A - B` | in A, **not** in B |
| Symmetric difference | `A.symmetric_difference(B)` | `A ^ B` | in **exactly one** |

Comparisons: `A.issubset(B)`, `A.issuperset(B)`, `A.isdisjoint(B)`, `A == B`

Full table: https://wesmckinney.com/book/python-builtin.html#tbl-table_set_operations

In [83]:
A = {1, 2, 3, 4, 5}
B = {4, 5, 6, 7}

print("Union               :", A | B)
print("Intersection        :", A & B)
print("Difference A - B    :", A - B)
print("Difference B - A    :", B - A)
print("Symmetric difference:", A ^ B)

print({1, 2}.issubset(A), A.issuperset({1, 2}), A.isdisjoint({10, 20}))
print({1, 2, 3} == {3, 2, 1})     # order does not matter

Union               : {1, 2, 3, 4, 5, 6, 7}
Intersection        : {4, 5}
Difference A - B    : {1, 2, 3}
Difference B - A    : {6, 7}
Symmetric difference: {1, 2, 3, 6, 7}
True True True
True


## Sets: useful patterns
* Remove duplicates: set(values)
* Are all values different? Compare len(values) with len(set(values))

In [84]:
readings = [3, 1, 3, 2, 1, 3]
print(sorted(set(readings)))          # unique and ordered

ids_ok = [101, 102, 103, 104]
ids_bad = [101, 102, 103, 102]
print(len(ids_ok) == len(set(ids_ok)))
print(len(ids_bad) == len(set(ids_bad)))

[1, 2, 3]
True
False


## Two courses, two registration lists (as they came from the registration form).

* How many different students registered for the Python course?
* Who takes both courses?
* Who takes Python but not Statistics?
* Who takes exactly one of the two courses?
* Does python_course contain a duplicate registration? Print True or False.


In [86]:
python_course = ["Anna", "Janis", "Marta", "Olegs", "Anna", "Liga"]
stats_course = ["Marta", "Peteris", "Liga", "Ilze", "Olegs"]

python_set = set(python_course)
stats_set = set(stats_course)

print("Different Python students:", len(python_set))

print("Both courses:", python_set & stats_set)

print("Python only:", python_set - stats_set)

print("Exactly one course:", python_set ^ stats_set)

has_duplicates = len(python_course) != len(python_set)
print("Duplicate registration:", has_duplicates)

Different Python students: 5
Both courses: {'Olegs', 'Liga', 'Marta'}
Python only: {'Anna', 'Janis'}
Exactly one course: {'Janis', 'Peteris', 'Ilze', 'Anna'}
Duplicate registration: True


## Dictionaries

A dictionary stores a collection of key-value pairs, where key and value are Python objects. Each key is associated with a value so that a value can be conveniently retrieved, inserted, modified, or deleted given a particular key.

In [ ]:
dict1 = {'Monday': 'Moon', 'Tuesday' : 'Mars', 'Wednesday' : 'Mercury', 'Thursday' : 'Jupiter',
         'Friday' : 'Venus','Saturday' : 'Saturn'}
print(dict1)
print(type(dict1))
print(dict1.keys())
print(dict1.values())

{'Monday': 'Moon', 'Tuesday': 'Mars', 'Wednesday': 'Mercury', 'Thursday': 'Jupiter', 'Friday': 'Venus', 'Saturday': 'Saturn'}
<class 'dict'>
dict_keys(['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'])
dict_values(['Moon', 'Mars', 'Mercury', 'Jupiter', 'Venus', 'Saturn'])


In [ ]:
dict1['Sunday'] = 'Sun'
print(dict1)

{'Monday': 'Moon', 'Tuesday': 'Mars', 'Wednesday': 'Mercury', 'Thursday': 'Jupiter', 'Friday': 'Venus', 'Saturday': 'Saturn', 'Sunday': 'Sun'}


In [ ]:
dict1.pop('Monday')
print(dict1)

{'Tuesday': 'Mars', 'Wednesday': 'Mercury', 'Thursday': 'Jupiter', 'Friday': 'Venus', 'Saturday': 'Saturn', 'Sunday': 'Sun'}


## Dictionaries: reading

| Expression | Result if the key is missing |
|---|---|
| `d[key]` | `KeyError` |
| `d.get(key)` | `None` |
| `d.get(key, default)` | `default` |
| `key in d` | `False` |

- `in` checks **keys**. For values: `x in d.values()`
- `len(d)` — the number of key-value pairs

In [89]:
capitals = {"Latvia": "Riga", "Estonia": "Tallinn", "Lithuania": "Vilnius"}

print(capitals["Latvia"])
print(len(capitals))

# ERROR: KeyError: 'Poland'
# print(capitals["Poland"])

print(capitals.get("Poland"))
print(capitals.get("Poland", "unknown"))
print("Poland" in capitals)
print("Riga" in capitals)                   # False - Riga is a value, not a key
print("Riga" in capitals.values())

Riga
3
None
unknown
False
False
True


## Dictionaries: adding, updating, deleting
* d[key] = value — a new key adds a pair, an existing key replaces the value
*  d.update(other) — several pairs at once
* d.pop(key) — removes the pair and returns its value; del d[key] — just removes

In [90]:
capitals["Finland"] = "Helsinki"       # new key
print(capitals)

capitals["Finland"] = "Helsingfors"    # existing key - the value is replaced
print(capitals)

capitals.update({"Sweden": "Stockholm", "Finland": "Helsinki"})
print(capitals)

removed = capitals.pop("Sweden")
print(removed)

del capitals["Finland"]
print(capitals)

{'Latvia': 'Riga', 'Estonia': 'Tallinn', 'Lithuania': 'Vilnius', 'Finland': 'Helsinki'}
{'Latvia': 'Riga', 'Estonia': 'Tallinn', 'Lithuania': 'Vilnius', 'Finland': 'Helsingfors'}
{'Latvia': 'Riga', 'Estonia': 'Tallinn', 'Lithuania': 'Vilnius', 'Finland': 'Helsinki', 'Sweden': 'Stockholm'}
Stockholm
{'Latvia': 'Riga', 'Estonia': 'Tallinn', 'Lithuania': 'Vilnius'}


### Complex keys

In [91]:
coords = dict()
coords[(56.9677, 24.1056)] = 'Riga'
coords[(51.5072, 0.12)] = 'London'
print(coords)

{(56.9677, 24.1056): 'Riga', (51.5072, 0.12): 'London'}


### Hashability

* A dictionary (and a set) finds elements through a hash — a number computed from the object's content
* The hash must never change → only immutable objects are hashable
* Hashable: int, float, str, bool, tuple — not hashable: list, set, dict

In [92]:
coords[[50.8476, 4.3572]] = 'Brussel'
print(coords)

TypeError: unhashable type: 'list'

In [ ]:
coords[{50.8476, 4.3572}] = 'Brussel'
print(coords)

TypeError: unhashable type: 'set'

In [93]:
hash((50.8476, 4.3572))

1658362060139021477

In [ ]:
hash({50.8476, 4.3572})

TypeError: unhashable type: 'set'

In [ ]:
hash([50.8476, 4.3572])


TypeError: unhashable type: 'list'

In [ ]:
coords[(50.8476, 4.3572)] = 'Brussel'
print(coords)

{(56.9677, 24.1056): 'Riga', (51.5072, 0.12): 'London', (50.8476, 4.3572): 'Brussel'}


## dictionaries -> Colab

prices is the menu of a cafe (EUR).

* Add "tea" for 2.20.
* The price of "coffee" goes up to 3.00 — update it.
* Print the price of "juice". If it is not on the menu, print "not on the menu" (no error!).
* Remove "cake" from the menu and print the price it had.
* Is anything sold for exactly 2.20? Print True or False.

In [100]:
prices = {"coffee": 2.80, "cake": 4.50, "sandwich": 5.20, "water": 1.50}


prices["tea"] = 2.20
print(prices)

prices["coffee"] = 3.00
print(prices)

print("Juice:", prices.get("juice", "not on the menu"))

cake_price = prices.pop("cake")
print("Cake price was:", cake_price)


print(2.20 in prices.values())

{'coffee': 2.8, 'cake': 4.5, 'sandwich': 5.2, 'water': 1.5, 'tea': 2.2}
{'coffee': 3.0, 'cake': 4.5, 'sandwich': 5.2, 'water': 1.5, 'tea': 2.2}
Juice: not on the menu
Cake price was: 4.5
True


##  Which structure to choose?

| Question | Structure |
|---|---|
| Do I look values up by a name / code / ID? | **dictionary** |
| Must every value be unique, or do I compare groups? | **set** |
| Is it a fixed record that never changes? | **tuple** |
| Otherwise: an ordered collection that changes | **list** |

## mini case: two bus routes -> Colab

*route_a* and *route_b* list the stops of two bus routes in travel order. boardings shows how many passengers boarded route A at each stop this morning.

* Print the number of stops on route A and its two terminals (first and last stop) as a tuple.
* Print the stops of route A for the return trip (reverse order).
* At which stops can a passenger transfer between the two routes?
* ow many different stops do the two routes serve together?
* Which stops are served only by route B? Print them in alphabetical order.
* How many passengers boarded route A in total? What is the average per stop (1 decimal)?
* Which stop was the busiest?
* Route A is extended: add the stop "Airport" to the end of the route with 12 boardings. Print the updated route as one string: Depot -> Market -> ...




In [101]:
route_a = ["Depot", "Market", "Central Station", "Old Town", "University", "Hospital"]
route_b = ["Harbour", "Old Town", "Central Station", "Stadium", "Zoo"]

boardings = {"Depot": 14, "Market": 37, "Central Station": 85,
             "Old Town": 52, "University": 61, "Hospital": 23}

## Summary
* List — ordered, mutable: add, remove, sort, slice
* Tuple — ordered, immutable: fixed records, unpacking, usable as a key
* Set — unique, unordered: de-duplication, union / intersection / difference
* Dictionary — key → value: fast look-up, get, keys / values / items

Next: 2.2 Control Flow — if, for, while

# Thank you